### DISCUSSION TITLE
In-Range Data Types (easier way than Method Contracts)

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#1198] [General]
> EDIT:
> So, there are 4 possible ways to do it:
> 1.	Method contracts. It is verbose but can write generic conditions.
> 2.	Constrained Data Types:  It is a compact syntax and can be extended to use regular expressions.
> 3.	Validation attributes: Like those used in ASP.NET MVC. It has no effect on the c# syntax if self, and we can use  parameters that indicate the error message or other actions. But using attribute for individual method parameters and return value can make the code long and ugly!
> 4.	Writing an analyzer, which I cannot exactly understand !
> 
> Sure we all want some syntax to simplify validation codes.
> We can make compromises. For example:
>  If the constrained types can cause problems, they can only be used to define properties and methods (parameters and return values). They are nit true types but just syntax annotations to tell compiler how validate values, so the compiler can convert then to contracts under the hood.
> Attributes also can be used with  properties and methods to supply error message and additional options to tell the compiler what to do when the data are not valid. 
> This approach says that contracts can still be used in more complicated situations (but I see that it is easer to use regular code statements)** 
> 

I suggest to improve current numeric types by adding a way to set a value range. I don't ask for numeric range array . I want a limited value range variable that can have a single value within the range. This is a sort of value validation, that can be applied to Properties and will affect binding them to data entry controls.
This can be done by adding two readonly properties to each data type: Min, Max. The compiler should check them before assigning any value to the object. 
Max and Min Values can also be set via the constructor.
The can be done also by some attributes, but it will be easier to give it a syntax like this
`[5,10]int i = 7;`
or 
`<5,10>int i = 7;`
The compiler must not allow i to have any value < 5 or >10. 
It should validate values before any assignment and raise overflow exceptions.
One can omit one limit:
```
 [,100]byte b = 90;
[13,]int a = 15;
```
or  
```
<,100>byte b = 90;
<13,>int a = 15;
```
This will make it easy to write auto-properties like this:
`[18,30]int Age {get; set;};`
Or
`<18,30>int Age {get; set;};`

In-Range Data Types give an easier way to achieve Method Contracts. It is very clear that this syntax 
```
public <0,Count>int Insert(T item, <0,Count>int index)   {
    return InsertCore(item, index);
}
```
is easear than:
```
public int Insert(T item, int index)
    requires index >= 0 && index <= Count
    ensures return >= 0 && return < Count
{
    return InsertCore(item, index);
}
```
If you constrained the data type, then you've already constrained all other language elements built with it.

Note:
If you want to complicate things, I suggest to use regular expression syntax such as:
`[reg exp]int x;`
This can allow writting constrains on string data types as well.
 
Another possible implementation:
```
    struct Range<T > where T : struct, IComparable
    {
        T Min { get; set; }
        T Max { get; set; }

        public Range(T min, T max)
        {
            Min = min;
            Max = max;
            _value = min;
        }

        T _value;
        public T Value
        {
            get => _value;

            set
            {
                if (value.CompareTo(Max) == 1 || value.CompareTo(Min) == -1)
                    throw new ArgumentOutOfRangeException();

                _value = value;
            }
        }

        public static implicit operator T(Range<T> r)
        {
            return r.Value;
        }
    }

```
Usage:
```
var range = new Range<int>(18, 60);
range.value = 20;
```

which can have a  short syntax:
```
<18, 60>int range;
range = 20;
```
The importance of this synatax appeares when declaring properties and method parameters and return values
For example:

```
<10, 90>int Foo(<1, 10>int x)
{
    // ........
    return result;
}
```

The compiler translates it to:

```
int Foo(int x)
{
  var x_range = new Range<int>(1, 10);
  x_range.Value = x;
  // ........
  var result_range = new Range<int>(10, 90);
  result_range.Value = result;
  return result;
}
```

**Comment by alrz** [comment]
https://github.com/dotnet/csharplang/issues/105

**Comment by Joe4evr** [comment]
And for *actual* ranges: dotnet/roslyn#23205

**Comment by ghost** [comment]
I don't ask for numeric range array . I want a limited value range variable. that can have a single value granteed to be in the range. This is a sort of value validation, that can be aplied to Properties and will affect binding them to data entry controls.

**Comment by yaakov-h** [comment]
#105?

**Comment by MkazemAkhgary** [comment]
I dont think its needed. You can already achieve it using struct and implicit operator.

```
struct Percent
{
    int value;
    public Percent(int value)
    {
        if(value > 100 || value < 0)
            throw new ArgumentOutOfRangeException(nameof(value), "percentage must be within 0-100");
        this.value = value;
    }
    
    public static implicit operator int(Percent value) => value.value;
    public static implicit operator Percent(int value) => new Percent(value);
}
```

**Comment by MkazemAkhgary** [comment]
In case default value should not be 0 use constant offset.

```
struct YoungAge 
{ 
    const int min = 18;
    const int max = 35;
    int age; 
    public YoungAge(int age) 
    { 
        if(age > max || age < min) 
          throw new ArgumentOutOfRangeException(nameof(age), $"young adult age is beween {min}-{max}"); 
        this.age = age - min;
    }
    public static implicit operator int(YoungAge age) => age.age + min; 
    public static implicit operator YoungAge(int age) => new YoungAge(age - min); 
}
```

**Comment by yaakov-h** [comment]
@MkazemAkhgary That gives you no compile-time safety, it only automates a handful of exceptions.

**Comment by Joe4evr** [comment]
> I don't ask for numeric range array. I want a limited value range variable.

Yes, but that is *constraining* a value to be *within* a range. *Completely* different from the range itself.

**Comment by ghost** [comment]
@MkazemAkhgary 
I think you miss the point. Almost every thing can be done with c# 3.0. The idea here is to save time and effort wasted in repeated tasks. We lived before Tuples but our lives will be easer after them.
C# is compiled to IL then IL is compiled to machine language. This 2-step compilation gives us the opportunity to complicate c# syntax as we want to save programmers time and effort, without affecting the JIT Compiler. 
So, let's do that. Let's add syntax to shorten repeated poring tasks. Why not?
You wrote a page to accomplish what I suggested in one line!
Thanks.

**Comment by ghost** [comment]
@yaakov-h 
Why you keep pasting the same link? 
I looked there put couldn't find something similar to my idea, so I posted it there. Can you please explane what you mean?

**Comment by ghost** [comment]
@Joe4evr 
I don't get it. If there a method to do what I suggested here in short easy way, please write a code sample.
Thanks.

**Comment by yaakov-h** [comment]
ah, I didn't see @alrz had posted it already.

@MohammadHamdyGhanem If you click through from that proposal to the original Roslyn issue, you should see how it's relevant.

**Comment by ghost** [comment]
@yaakov-h 
Sorry, I cann't relly understand you. Could you care to explain. "see how it's relevant". What is "it"? Are we in the same page here?

**Comment by yaakov-h** [comment]
I fear you've fallen out of the book completely. 😄 

You appear to asking for a way to constrain a variable or property to have maximum and minimum values.

This sounds like part of [method contracts](https://github.com/dotnet/csharplang/issues/105), which is [a proposal](https://github.com/dotnet/roslyn/issues/119) based upon the old Microsoft Research project, Code Contracts.

The idea here is more general than your specific case, but would allow you do something like:

```cs
class RiskyDriver
{
    public int Age ensures return >= 18 && return < 25 { get; }
}
```

**Comment by ghost** [comment]
Yes. I read that and it is very clear that my syntax is easear:
```
public <0,Count>int Insert(T item, <0,Count>int index)   {
    return InsertCore(item, index);
}
```
than
```
public int Insert(T item, int index)
    requires index >= 0 && index <= Count
    ensures return >= 0 && return < Count
{
    return InsertCore(item, index);
}
```
If you constrained the data type, then you've already constrained all other language elements built with it.

**Comment by jveselka** [comment]
@MohammadHamdyGhanem 

> That gives you no compile-time safety, it only automates a handful of exceptions.

And how would you ensure compile-time safety in cases like this one
`<,100>int i = new Random().Next(200);`
Are you ok with compile-time safety only for constants? Seems like a very minor benefit to me.

Regarding method contracts, your syntax is indeed shorter, but far less powerfull. (Only limited to buildin numeric type limits.) If you ask me, I would rather take extra features over fewer characters.

**Comment by ghost** [comment]
@zippec 
If it is implemented as a real c# syntax, sure the compiler will validate values before any assignment and raise overflow exceptions. This is what I expect and asdek for when I said "
<5,10>int i = 7;
The compiler must not allow i to have any value < 5 or >10. "

**Comment by Joe4evr** [comment]
> If you ask me, I would rather take extra features over fewer characters.

This is the most important thing to keep in mind. Expressiveness >>>>>>>>>>>>>>>>>>>>> lines/characters of code.

For example, I have this ternary expression that assigns a value of a 3-state enum depending on whether an entry exists in a dictionary and if that happens to be equal to a pre-determined value:
```cs
E = dictionary.TryGetValue(key, out var name)
    ? (name == OtherName ? E.Foo : E.Bar)
    : E.Baz;
```
[Once `switch` expressions are released](https://github.com/dotnet/csharplang/issues/1054#issuecomment-348009587) that can become:
```cs
E = dictionary.TryGetValue(key, out var name) switch
{
    true when name == OtherName => E.Foo,
    true  => E.Bar,
    false => E.Baz
};
```
which is twice the lines of code, but infinitely more clearer in its intent.

**Comment by ghost** [comment]
Syntax that is difficult to write and understand will be ignored by programmers.
Validation attributes already exist in asp.NET MVC and it should be implemented in c# itself at the compiler level to be a part of data types features. If you consider the regular expressions validarion,this will be more powerful than contracts.

**Comment by DavidArno** [comment]
Duplicate of #846.

And, [as I said on that thread, this is a bad idea in my view](https://github.com/dotnet/csharplang/issues/846#issuecomment-324856872):
> Having used ranged integer types in Ada, back in the day, I can say with confidence that such types are way more trouble than they are worth. Sure it gives built in range checking, but it then leads to a plethora of incompatible int types throughout the code that have to be explicitly cast one from the other.

> Also, with this proposal, the ranges are being expressed with "magic numbers", directly at the point of use, this would create a maintenance headache if those ranges had to change.

> So a definite 👎 from me.

**Comment by ghost** [comment]
@DavidArno 
This is not a new data type and doesn't require any different casting approach.  It is only a run time constraint that makes us write less if statements.
Yes, the two topics are related. I did not saw that old one, but mine here is the generic one.
Changing range wont be a problem if we use constants. This is what we do now with regular if statements and what we will do with method contracts and in range types shall any be introduced.

**Comment by ghost** [comment]
It appears that this proposal is repeatedly posted.
#413 
#846 
Sure it has some popularity.

**Comment by tannergooding** [comment]
One of the primary issues with contracts, in my opinion, is that relying on compile time validation is very limited. 

In my experience, compile time validation works great for constants, but not so well for arbitrary inputs. For real world applications, data is generally arbitrary input coming from the user in the form of a string, which must be validated, processed, and converted to the appropriate type before it can be stored in some database.

Because of this, compile time validation doesn't actually catch a lot of the errors that will crop up in production. It really can only catch the simplest of errors and will still get converted to an `if (contract.fail) throwʼ expression under the hood to catch the rest.

For the cases that compile time validation can catch, an analyzer will work just fine. As such, what I would want to see (instead of this or method contracts) is:
1. An easy way to write boilerplate code (there is the code generator proposal for this, but most Ides provide some functionality to make this easier as well).
2. Some analyzer templates for input validation to make this task even easier
3. The ability to indicate that a library has a compile time analyzer the compiler should expect to be present (unless explicitly opted out) so that library authors can enforce certain behaviors or validation like this (today, we don't have an easy way to say that an analyzer exists for a library and that it should be considered opt-out by default)

**Comment by iam3yal** [comment]
@tannergooding That's not what DbC is about but preconditions, postconditions and invariants, data validation has nothing to do with it at all.

**Comment by Pzixel** [comment]
@MohammadHamdyGhanem 

```csharp
struct Age 
{ 
    const int min = 18;
    const int max = 35;
    public int Age { get; }
    public Age(int age) 
       requires age >= min && age <= max
    {
        this.age = age;
    }
}
```

Now you constrained all clients of your `Age`.

**Comment by tannergooding** [comment]
They all essentially fall down into data validation.

Preconditions either statically (at compile time) or dynamically (at runtime) enforce that the inputs or some state match the required values.

Postconditons (ensures) do the same thing for outputs or state. 

In both cases, it is essentially syntactic sugar for an `if (!cond) throw`, but one that the compiler can also take advantage of (which an analyzer could also do).

Invariants are essentially compiler hints or assert/static_asserts to ensure some condition is met between the pre/post conditions.

If you have them across the board, they can help catch a number of errors, but I think analyzers can achieve much of the same functionality and would rather see several other features prioritized first.

**Comment by ghost** [comment]
So, there are 4 possible ways to do it:
1.	Method contracts. It is verbose but can write generic conditions.
2.	Constrained Data Types:  It is a compact syntax and can be extended to use regular expressions.
3.	Validation attributes: Like those used in ASP.NET MVC. It has no effect on the c# syntax if self, and we can use  parameters that indicate the error message or other actions. But using attribute for individual method parameters and return value can make the code long and ugly!
4.	Writing an analyzer, which I cannot exactly understand !

Sure we all want some syntax to simplify validation codes.
We can make compromises. For example:
 If the constrained types can cause problems, they can only be used to define properties and methods (parameters and return values). They are nit true types but just syntax annotations to tell compiler how validate values, so the compiler can convert then to contracts under the hood.
Attributes also can be used with  properties and methods to supply error message and additional options to tell the compiler what to do when the data are not valid. 
This approach says that contracts can still be used in more complicated situations (but I see that it is easer to use regular code statements)

**Comment by iam3yal** [comment]
@tannergooding It doesn't matter where things happen and whether there are overlaps, DbC is a tool used by the programmer to know whether all the invariants in the system are met, many times these conditions are removed from release builds for performance's sake whereas data validation is kept during release builds.

What I'm saying is there's obviously some overlaps but they are not the same, they are used for different purposes so maybe I exaggerated by saying that it has nothing to do with data validation but really they solve two different things.

I used CodeContract long time ago and iirc it was executing the code in the background in order to invalidate the contracts, there's much more to it and yes you could allow contracts only on public APIs to validate the input but that was a feature not the goal.

**Edit:** One more thing, all contracts needs to be visible whereas for data validation it isn't a requirement.

**Edit2:** Rehprased. 😆

**Comment by Pzixel** [comment]
@eyalsk data validation as contract would be really valuable for me. Fail early is the best practice, so fail at compile time with "you pass 10 here, but it must be 0 to 8" would be really good. It's not working with dynamic data, but it still can generate guard methods that validates contracts at runtime. So you just write declarative `x >= 5` and compiler emit all required `throw OutOfRangeException`. We don't have to write all this prelude of `throw` in every method. As a bonus, if we have some public method and some private method and we validates the same arguments in two places, we can omit checks in one of them. For example, it's quite common in LINQ to check arguments for null in all places, even in private ones. I'm not sure it worth to implement such a logic, however, still being possible.

**Comment by iam3yal** [comment]
@Pzixel 

> but it still can generate guard methods that validates contracts at runtime

That's exactly why I wrote you shouldn't mix them up... because if anything this should be a separate feature.

**Comment by Pzixel** [comment]
@eyalsk why? They really seem to be coupled.

**Comment by iam3yal** [comment]
@Pzixel You may want to read [this](http://enterprisecraftsmanship.com/2015/02/14/code-contracts-vs-input-validation/).

**Comment by bondsbw** [comment]
I favor `int<1,10>` over `<1, 10>int` simply because the former is a familiar way to provide constraints.

**Comment by bondsbw** [comment]
@tannergooding 
> Because of this, compile time validation doesn't actually catch a lot of the errors that will crop up in production. It really can only catch the simplest of errors and will still get converted to an `if (contract.fail) throwʼ expression under the hood to catch the rest.

But it gets converted to that _everywhere_.  I don't have to write that (or method contract equivalent) in all the places such data may flow.

**Comment by bondsbw** [comment]
@DavidArno
> > Having used ranged integer types in Ada, back in the day, I can say with confidence that such types are way more trouble than they are worth. Sure it gives built in range checking, but it then leads to a plethora of incompatible int types throughout the code that have to be explicitly cast one from the other.

If we didn't already have unconstrained types I might agree with you.  

If I opt into using constraints, then I obviously want the compiler to give me those warnings.  I want conversion to be a headache because I'm probably doing something wrong.

> > Also, with this proposal, the ranges are being expressed with "magic numbers", directly at the point of use, this would create a maintenance headache if those ranges had to change.

Just like any use of magic numbers, feel free to use constants.  And while you're at it, combine them with [type aliases](https://github.com/dotnet/csharplang/issues/846#issuecomment-324758832).

**Comment by Pzixel** [comment]
@bondsbw 

> But it gets converted to that everywhere. I don't have to write that (or method contract equivalent) in all the places such data may flow.

I think data contracts may be inferred. For example above

```csharp
struct Age 
{ 
    const int min = 18;
    const int max = 35;
    public int Age { get; }
    public Age(int age) 
       requires age >= min && age <= max
    {
        this.age = age;
    }

    public static Age Create(int age) => new Age(age);
}
```

I'd like `Create` to infer `requires age >= min && age <= max` do you don't have to type it manually.
However, I believe that it may be impossible due to very intensive compiler work required.

**Comment by bondsbw** [comment]
@Pzixel I don't really buy that constrained numeric types requires "intense" compiler work*.  It is just a way to define generic numeric types and enforce a different set of constraints at compile time.  Mostly the differences from the perspective of the compiler are MinValue and MaxValue.

The real work will be in library support, so that generic/constrained numeric values are accepted in places that unconstrained values are currently allowed.

*_Relatively speaking, that is.  Of course just changing the simplest thing can take a huge amount of effort._

**Comment by Pzixel** [comment]
@bondsbw I didn't say that it's your feature that required intense work. Please, reread it if it's not clear.

**Comment by bondsbw** [comment]
@Pzixel To do that, you'll either need a type annotation of some kind (constrained types) or deep static analysis.  I'm not sure what the latter buys over the former.

**Comment by Pzixel** [comment]
@bondsbw deep static analysis is indeed much smarter and better, however it may be laggy, it's what I said before. About constrained types - I like the idea of typing, but specifying it manually may be painful. Why we introduce `int<1,10>`, but we don't do it for `int<oddNumber>`, or `int<isPrime>`? They are useful too. For example it could be much better to use the latter in hashset/dictionary implementation.

I often encounter ranges, but they are always dynamically sized. You don't have int<1,10> in most cases, you have `int<a,b>` where `a` and `b` is user input/config values/database values/... They cannot be figured out by the compiler.

**Comment by iam3yal** [comment]
I think that it's better to discuss DbC at one place which is at #105 as this covers more than just what this proposal aims to achieve and unlike this one it's also championed.

**Comment by ghost** [comment]
One advantage of using constrained types, is that the caller can simply notice them in member info tooltips in a glimpse. Method Contracts isn't that easily readable in a tooltip. so, constained types are self documented.

**Comment by yaakov-h** [comment]
Code Contracts had a Visual Studio extension that included the contracts in intellisense tooltips.

**Comment by ghost** [comment]
Constrained Types also can be used with arrays and lists:
```
<5,10>int[] x;
List< <5,10>int > y;
List< <5,10>int[] > z;
```
You can't do this with a contract. And when you write a contract for a method parameter of type List<int>, you'll use LinQ to check each element. 
So, each feature has its unique capabilities.

**Comment by yaakov-h** [comment]
You can do this today with Code Contracts:

`Contract.ForAll(someEnumerable, x => x >= 5 && x < 10);`

I assume language-level contracts would have a similar construct.

**Comment by ghost** [comment]
> Code Contracts had a Visual Studio extension that included the contracts in intellisense tooltips.

Yes, but the size of the tip can be huge and not easy to read.

**Comment by yaakov-h** [comment]
You're arguing a language-level feature based on your own assumptions of how the IDE will represent it to the user?

**Comment by ghost** [comment]
> You can do this today with Code Contracts:
> Contract.ForAll(someEnumerable, x => x >= 5 && x < 10);

this is only a contract for parameters and return values only? Right?
If you declare a list as a field, and used 20 methods and properties to deal with it, you will need to write contracts in 20 places, while you use a constrained type to declare the List once!

**Comment by yaakov-h** [comment]
No, also properties and fields. That's what invariants are for.

**Comment by bondsbw** [comment]
@Pzixel
> Why we introduce `int<1,10>`, but we don't do it for `int<oddNumber>`, or `int<isPrime>`? They are useful too.

No doubt, those would be very powerful.

The former is logically a tweak of the existing types, where MinValue and MaxValue are modifiable constants.

I think the constraint functions you suggested are very interesting.  But they represent a lot more customization than tweaking existing constraints and would require your code to run in the compiler at build time and could do things that cause the compiler or IDE to hang or crash.  That said, analyzers face the same dilemma, so there is some precedent in C#.

**Comment by MkazemAkhgary** [comment]
@bondsbw

>No doubt, those would be very powerful.

Are you sure? Making c# so complicated and turn it into a very non procedual language just to say odd or even numbers? How much you ever came to this scenarios?

I say its not powerful at all. But you can do pretty much every thing with current power of c#.

**Comment by Pzixel** [comment]
@bondsbw 

> The former is logically a tweak of the existing types, where MinValue and MaxValue are modifiable constants.

It's the same. With ranges we check if int is in some range, with `int<isPrime>` compiler perform same check, but verify that number is actually a prime number. No difference at all.

> I say its not powerful at all. But you can do pretty much every thing with current power of c#.

You actually can do _everything_ since C# is Turing complete language.

**Comment by bondsbw** [comment]
@Pzixel It is not the same.  Constants do not introduce the ability to hang the compiler:

```c#
bool isPrime(int i) {
    while (true) {}
    return true;
}
```

**Comment by bondsbw** [comment]
@MkazemAkhgary
> How much you ever came to this scenarios?

Every time I use enums.  Every time I need to limit input to something like 0-100.  Surveys where the only valid values are 1, 2, 3, 4, and 5.  Array indexing.

Is it _possible_ to write applications that check these constraints today?  Sure.  Is it _possible_ to pass strings around using the `object` data type?  Of course.  But you usually don't, because you want type safety.  This suggestion enhances type safety of numeric data types.

**Comment by Pzixel** [comment]
@bondsbw you have to have built in compiler support for `isPrime` just like you want it to check range. We have Roslyn that has its internal `isPrime` function that is written correctly which checks if input integer is valid prime number. How check if number is prime differs from checking if it's in range?

> Every time I use enums. Every time I need to limit input to something like 0-100. Surveys where the only valid values are 1, 2, 3, 4, and 5. Array indexing.

Array indexing is based on `array.Length` which is unknown at compile time for 99% cases.

**Comment by bondsbw** [comment]
@Pzixel Built-in functions work but the usefulness diminishes.  Might as well have `primeint` and `oddint` if you can't customize the logic.

Fixed array sizes are very common.

**Comment by Pzixel** [comment]
> Fixed array sizes are very common.

Can you proove it? Common cases with fixed size arrays are reading string using some buffer and so on. I don't believe we need a feature for locals only, types are useful when crossing method boundaries only.

**Comment by bondsbw** [comment]
@Pzixel No, you made the claim that 99% of arrays have sizes that cannot be determined until runtime.  The burden is yours to prove your claim.  Regardless, you are focusing on a single case of several I described.

Who said this was for locals only?  You are the first to mention it.

**Comment by Pzixel** [comment]
@bondsbw 

> No, you made the claim that 99% of arrays have sizes that cannot be determined until runtime. The burden is yours to prove your claim. Regardless, you are focusing on a single case of several I described.

Ok, I will after you prove yours :)

> Who said this was for locals only? You are the first to mention it.

You don't know array length when you passed it into the method

`public void Foo(params int[] args) => // do we now length of args at compile time here?`

**Comment by bondsbw** [comment]
That isn't how burden of proof works.  You made the claim.  I made the counterclaim.  You must prove the claim before expecting me to prove the counterclaim.

You do know the array length when it is a constant.  Obviously if it isn't a constant, then this feature doesn't apply.

**Comment by theunrepentantgeek** [comment]
There's another reason why Range data types are a bad idea - and it's something that I haven't seen mentioned yet in this thread. Consider this class and imagine you have a non-trivial system (>250kloc) that makes extensive use of it:

```
public class AdultWhoCanDrink
{
    public string FullName { get; }
    public string KnownAs { get; }
    public string OrderKey { get; }
    public <18,99>int Age { get; }
}
```

What happens when the local government *changes the law* and now the legal age for drinking alcohol is going to change to 21 years of age?

Worse, the type declaration `<18,99> int` is dispersed throughout your codebase - and the new legislation becomes effective in 28 days. 

If you think this is a contrived example, I challenge you to check out some of the so-called *constants* in your local jurisdiction. Things like age thresholds (e.g. for voting, drinking age, or for military service), sales tax rates (e.g. NZ GST), and food standards (e.g. how much meat is needed a sausage) are inevitably political decisions - and those can change, sometimes with very short notice.

Simply using `int` throughout the code isn't a great solution, suffering as it does from *primitive obsession*.

Instead, we can define `DrinkingAge` with the appropriate rules and use that throughout the code. Then, when changes are required, we only need to make the change in one place.

We can already define `DrinkingAge` as a struct, incurring a very low memory cost (lower than a generic range type, in fact). 

If the language **did** support range types, we'd only need to use them in one place … so the benefit to the codebase would be very low.

A last couple of comments …

… with Method Contracts, we could appropriately annotate the `DrinkingAge` type and get enforcement throughout the codebase without needing to repeat the rules over and over. 

… if memory serves (it's been a long while since comparing programming languages in my university days), the experience of Ada developers using ranged type reflects this problem as well. They have nice characteristics when first writing code, but at the cost of making future feature development significantly harder.

**Comment by bondsbw** [comment]
@theunrepentantgeek
> What happens when the local government changes the law and now the legal age for drinking alcohol is going to change to 21 years of age?

If such changes occur regularly, or if you need to deal with multiple jurisdictions that have different definitions, then keep it as an `int` or use a type which checks against a variable range, say from a database or configuration file.  Those solutions fit your problem better.

Nobody is getting rid of `int`.

> If you think this is a contrived example, I challenge you to check out some of the so-called _constants_ in your local jurisdiction.

This is a design decision.  In a previous job we made the decision to hard-code the list of US states into an `enum` within an offline app, instead of creating a web service for periodically syncing the list.

Could the states change?  Sure, there has been talk about adding Puerto Rico for a while now.  Did we feel it was worth the cost instead of simply releasing an update to the app?  No, since the lead time for this change would be longer than our typical update schedule.

**Comment by bondsbw** [comment]
> Worse, the type declaration `<18,99> int` is dispersed throughout your codebase - and the new legislation becomes effective in 28 days.

As I mentioned before, [type aliases](https://github.com/dotnet/csharplang/issues/846#issuecomment-324758832) can be handy here:

```c#
type DrinkingAge = int<18,99>;
```

Now just change that once and push out an update.

_sorry, I really don't like putting the brackets and numbers first_

**Comment by ghost** [comment]
Another possible implementation:
```
    struct Range<T > where T : struct, IComparable
    {
        T Min { get; set; }
        T Max { get; set; }

        public Range(T min, T max)
        {
            Min = min;
            Max = max;
            _value = min;
        }

        T _value;
        public T Value
        {
            get => _value;

            set
            {
                if (value.CompareTo(Max) == 1 || value.CompareTo(Min) == -1)
                    throw new ArgumentOutOfRangeException();

                _value = value;
            }
        }

        public static implicit operator T(Range<T> r)
        {
            return r.Value;
        }
    }

```
Usage:
```
var range = new Range<int>(18, 60);
range.value = 20;
```

which can have a  short syntax:
```
<18, 60>int range;
range = 20;
```
The importance of this synatax appeares when declaring properties and method parameters and return values.

**Comment by Pzixel** [comment]
Anyway it has type `Range<int>` rather than `<18,60> int`. It means that this value may be passed everywhere `Range<int>` can be accepted. You probably want compiler to generate checks on `min` and `max` values in all methods, but this is no-no imho.

**Comment by ghost** [comment]
> You probably want compiler to generate checks on min and max values in all methods, but this is no-no imho

Why? checks exists already in the Value property. Whenever T is passed to a range<T>, the compiler  sets Range.Value = T and it will accept the value or throw an exception. 
On the oher hand, Rang<T> can be safly converted to T becuase it is a widening cast. I see this is an easy solution.

**Comment by Pzixel** [comment]
@MohammadHamdyGhanem 

```csharp
void Foo(<20, 40>int range) => Console.WriteLine(range);
<18, 60>int range = 59;
Foo(range);
```

Compiles to:
```csharp
void Foo(new Range<int> range) => Console.WriteLine(range.Value);
var range = new Range<int>(18, 60) { Value = 59 };
Foo(range);
```

And fails at runtime. Where is the profit from having this type?
  
The problem here is that we are not in C++ world and we cannot write `Range<int, 20, 40>`. Range tresholds are not encoded in the type.

**Comment by ghost** [comment]
> Where is the profit from having this type?

Look at this:

```
<10, 90>int Foo(<1, 10>int x)
{
    // ........
    return result;
}
```

The compiler translates it to:

```
int Foo(int x)
{
  var x_range = new Range<int>(1, 10);
  x_range.Value = x;
  // ........
  var result_range = new Range<int>(10, 90);
  result_range.Value = result;
  return result;
}
```

The the profit:
1- A compact time-saving code.
2- You easily realize what are method cntracts when you see this in a tooltip:
<10, 90>int Foo(<1, 10>int x)

> And fails at runtime

You miss the point of code validation. Most methods and properties throw exceptions even those of the coreFX. We do this not to make the code fail at rune-time, put to foce our logic on any programmer that uses our code, so he had to provide valid values, and if he is not sure, he uses the try catch structure. This is how you design a reusable code for yourself and your team. In all cases the Foo method will reais an exception if it gets invalid values. The deference is that we don't have to write long validating code over and over.

**Comment by Pzixel** [comment]
1.
> You miss the point of code validation. Most methods and properties throw exceptions even those of the coreFX. 

If I don't have compile time check I don't want to have it in the language. Ranges are rare enough. I'd rather prefer correct datetime interop (see NodaTime) than this one. Just write your own Range type then, you don't need language support for it.

2. 

> Look at this:

It's beside the point. I show you that this feature gives nothing to me. It doesn't help to pass invalid range somewhere. It just save me some typing when I can write `new <1,10>int(5)` instead of `new Range<int>(1,10) { Value = 5 }`.

**Comment by ghost** [comment]
You can't use "new Range<int>(1,10) { Value = 5 }" as a parameter type, or as a retrun value type, or as a property type. This is what the example shows.

**Comment by ghost** [comment]
>  It doesn't help to pass invalid range somewhere

Sure it does. I expect that the compiler will give you a design time error at:
```
<18, 60>int range;
range = 10;
```
and will give you a runtime error when a wrong value can't be calculated at design time. This is the standerd behavior with basic numeric types:
```
byte b = -1; // design time error
b = Foo( ) // if foo returns 300, runtime error
```

**Comment by HaloFour** [comment]
Frankly, the proposed `<min,max>int` syntax is pretty hideous and doesn't at all convey that it's related to contracts.  The suggestion that it compile down to some generic `Range` type doesn't demonstrate how that can be helpful at compile time to avoid violation and runtime errors.  That's on top of the fact that `Range` is going to mean something different in C# fairly soon.

If the language were to go down the path of exploring constrained types I'd much rather they be defined as proper type aliases with constraints, something akin to:

```cs
public type CustomerName : string;
public type CustomerId : Guid;
public type AgeRange : int where (AgeRange > 0 && AgeRange < 150);
```

The implementation could be a simple struct wrapper around the "extended" type:

```cs
public struct CustomerName {
    private readonly string value;

    private CustomerName(string value) => this.value = value;
    public static implicit operator string(CustomerName rhs) => rhs.value; 
    public static explicit operator CustomerName(string rhs) => new CustomerName(rhs);
}
```

Then you'd be able to use the type alias like a normal type and the compiler would require explicit casting to that type which would/should at least cause the developer enough pause to think about whether the type is valid.

When constraints are included those constraints could be embedded in the constructor to throw if violated at runtime.  They could also be encoded into attributes which would allow for compile time validation, although that could be tricky if the syntax for defining constraints is fairly open-ended.
